app/models/concerns/sql_functions.rb

```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+# Model to create SQL functions for common use
+module SqlFunctions
+  extend ActiveSupport::Concern
+
+  class_methods do
+    def add_valid_date_function
+      is_valid_date_function = <<-SQL.squish
+        create or replace function is_valid_date(str varchar)
```

Comment:
I really like the approach! To be honest, I'm not a SQL wizard so I'm not entirely sure of the semantics here. Is there a reason that this is called on demand instead of a migration? As I read it, every time `Fact.birthdays_in_range` is called, the function is added or replaced in the database. To me, intuitively, it would seem like it would be best to do it once instead of every time? (although I'm sure the performance impact is negligible)

As I'm writing this out, I'm realizing that there is an advantage by doing it this way in that you can modify the function whenever and not have to create a new migration... but I'm still interested to hear what you think!

Reply:
The standard Ruby schema format doesn't support defining SQL functions unfortunately so you could end up losing the function entirely in certain conditions. The only way to handle this with a migration would be to convert the schema to a raw SQL schema - which is an option if you prefer that, I just thought I should implement this with minimal disruption to the existing set up. (Also, I always try and work within Ruby schema where possible, it's much more readable when you're scanning the file to find DB fields, SQL schema can end up being a bit of a mess.)

There's advantages and disadvantages to either way, I wouldn't strongly say that one is better than the other, it just comes down to your preference.